File: src/lib/protocol/event-references.ts
import { decodeEntity } from './nip19';
import { kinds } from './kinds';
import { replyParent, replyRoot } from './tags';
import { isEventId, parseNostrEvent, type NostrEvent } from './event';
import { verifyEvent } from './event-verify';

export type EventReferenceKind =
  | 'reply-root'
  | 'reply-parent'
  | 'quote'
  | 'repost'
  | 'reaction'
  | 'deletion'
  | 'nostr-event';

export type EventReference = {
  readonly kind: EventReferenceKind;
  readonly id: string;
  readonly relays?: readonly string[];
  readonly authorPubkey?: string;
  readonly marker?: string;
  readonly source?: 'e' | 'q' | 'content' | 'repost' | 'reaction' | 'deletion';
};

export function eventReferences(event: NostrEvent): EventReference[] {
  const refs: EventReference[] = [];
  if (event.kind === kinds.repost) {
    pushEventTag(refs, 'repost', event.tags.filter(isETag).at(-1), 'repost');
    return dedupe(refs);
  }
  if (event.kind === kinds.genericRepost) {
    pushEventTag(refs, 'repost', event.tags.filter(isETag).at(-1), 'repost');
    return dedupe(refs);
  }
  if (event.kind === kinds.reaction) {
    pushEventTag(
      refs,
      'reaction',
      event.tags.filter(isETag).at(-1),
      'reaction',
    );
    return dedupe(refs);
  }
  if (event.kind === kinds.deletion) {
    for (const tag of event.tags.filter(isETag))
      pushEventTag(refs, 'deletion', tag, 'deletion');
    return dedupe(refs);
  }
  for (const tag of event.tags.filter(isETag)) {
    if (tag[3] === 'root') pushEventTag(refs, 'reply-root', tag, 'e');
    if (tag[3] === 'reply') pushEventTag(refs, 'reply-parent', tag, 'e');
  }
  if (!refs.some((ref) => ref.kind === 'reply-root'))
    push(refs, 'reply-root', replyRoot(event), { source: 'e' });
  if (!refs.some((ref) => ref.kind === 'reply-parent'))
    push(refs, 'reply-parent', replyParent(event), { source: 'e' });
  for (const tag of event.tags.filter((tag) => tag[0] === 'q'))
    push(refs, 'quote', tag[1], {
      relays: tag[2] ? [tag[2]] : [],
      source: 'q',
    });
  refs.push(...nostrEventReferences(event.content));
  return dedupe(refs);
}

export function verifiedNestedRepost(
  event: NostrEvent,
): NostrEvent | undefined {
  const repost =
    event.kind === kinds.repost || event.kind === kinds.genericRepost;
  if (!repost || !event.content.trim()) return undefined;
  try {
    const parsed = parseNostrEvent(JSON.parse(event.content));
    if (!parsed.ok) return undefined;
    const verified = verifyEvent(parsed.event);
    return verified.ok && repostTargetMatchesSource(event, verified.event)
      ? verified.event
      : undefined;
  } catch {
    return undefined;
  }
}

function repostTargetMatchesSource(
  source: NostrEvent,
  target: NostrEvent,
): boolean {
  if (source.kind === kinds.repost)
    return (
      target.kind === kinds.textNote && lastEventTagId(source) === target.id
    );
  if (source.kind === kinds.genericRepost) {
    const declared = lastEventTagId(source);
    return !declared || declared === target.id;
  }
  return false;
}

function lastEventTagId(event: NostrEvent): string | undefined {
  const id = event.tags.filter(isETag).at(-1)?.[1];
  return id && isEventId(id) ? id : undefined;
}

function nostrEventReferences(content: string): EventReference[] {
  return [...content.matchAll(/\bnostr:([a-z0-9]+)/gi)].flatMap((match) => {
    const decoded = decodeEntity(match[1] ?? '');
    if (!decoded) return [];
    if (decoded.type === 'note' && isEventId(decoded.data))
      return [{ kind: 'nostr-event', id: decoded.data, source: 'content' }];
    if (decoded.type === 'nevent' && isEventId(decoded.data.id))
      return [
        {
          kind: 'nostr-event',
          id: decoded.data.id,
          relays: decoded.data.relays ?? [],
          authorPubkey: decoded.data.author,
          source: 'content',
        },
      ];
    return [];
  });
}

function push(
  refs: EventReference[],
  kind: EventReferenceKind,
  id: string | undefined,
  patch: Partial<EventReference> = {},
): void {
  if (id && isEventId(id)) refs.push(cleanReference({ kind, id, ...patch }));
}

function pushEventTag(
  refs: EventReference[],
  kind: EventReferenceKind,
  tag: readonly string[] | undefined,
  source: EventReference['source'],
): void {
  push(refs, kind, tag?.[1], {
    relays: tag?.[2] ? [tag[2]] : [],
    marker: tag?.[3],
    authorPubkey: tag?.[4],
    source,
  });
}

function dedupe(refs: readonly EventReference[]): EventReference[] {
  const byId = new Map<string, EventReference>();
  for (const ref of refs) {
    const existing = byId.get(ref.id);
    byId.set(ref.id, existing ? mergeRef(existing, ref) : ref);
  }
  return [...byId.values()];
}

function mergeRef(a: EventReference, b: EventReference): EventReference {
  return cleanReference({
    ...a,
    relays: [...new Set([...(a.relays ?? []), ...(b.relays ?? [])])],
    authorPubkey: a.authorPubkey ?? b.authorPubkey,
    marker: a.marker ?? b.marker,
  });
}

function isETag(tag: readonly string[]): boolean {
  return tag[0] === 'e';
}

function cleanReference(reference: EventReference): EventReference {
  return Object.fromEntries(
    Object.entries(reference).filter(([, value]) =>
      Array.isArray(value) ? value.length > 0 : value !== undefined,
    ),
  ) as EventReference;
}

File: crates/lkjstr-protocol/src/event_references.rs
use crate::{
    EventReference, EventReferenceKind, EventReferenceSource, KIND_DELETION, KIND_GENERIC_REPOST,
    KIND_REACTION, KIND_REPOST, NostrEntity, NostrEvent, decode_nip19,
    event_reference_parts::{
        dedupe, last_e_tag, normalized_relays, push, push_event_tag, tag_name_is, tag_relay,
    },
    event_reference_scan::{nostr_entities, nostr_entity_spans},
    is_event_id, reply_parent, reply_root,
};

#[must_use]
pub fn event_references(event: &NostrEvent) -> Vec<EventReference> {
    let mut refs = Vec::new();
    match event.kind {
        KIND_REPOST | KIND_GENERIC_REPOST => {
            push_event_tag(
                &mut refs,
                EventReferenceKind::Repost,
                last_e_tag(event),
                EventReferenceSource::Repost,
            );
            return dedupe(refs);
        }
        KIND_REACTION => {
            push_event_tag(
                &mut refs,
                EventReferenceKind::Reaction,
                last_e_tag(event),
                EventReferenceSource::Reaction,
            );
            return dedupe(refs);
        }
        KIND_DELETION => {
            for tag in event.tags.iter().filter(|tag| tag_name_is(tag, "e")) {
                push_event_tag(
                    &mut refs,
                    EventReferenceKind::Deletion,
                    Some(tag.as_slice()),
                    EventReferenceSource::Deletion,
                );
            }
            return dedupe(refs);
        }
        _ => {}
    }
    push_marked_replies(&mut refs, event);
    push_missing_reply_fallbacks(&mut refs, event);
    for tag in event.tags.iter().filter(|tag| tag_name_is(tag, "q")) {
        push(
            &mut refs,
            EventReferenceKind::Quote,
            tag.get(1),
            EventReferenceSource::Q,
            tag_relay(tag),
        );
    }
    refs.extend(nostr_event_references(&event.content));
    dedupe(refs)
}

#[must_use]
pub fn strip_event_reference_tokens(content: &str, refs: &[EventReference]) -> String {
    if refs.is_empty() {
        return content.to_owned();
    }
    let mut output = String::new();
    let mut cursor = 0;
    let mut removed = false;
    for span in nostr_entity_spans(content) {
        let Some(id) = event_entity_id(&span.entity) else {
            continue;
        };
        if !refs.iter().any(|item| item.id == id) {
            continue;
        }
        output.push_str(&content[cursor..span.start]);
        cursor = span.end;
        removed = true;
    }
    if !removed {
        return content.to_owned();
    }
    output.push_str(&content[cursor..]);
    output.trim().to_owned()
}

fn push_marked_replies(refs: &mut Vec<EventReference>, event: &NostrEvent) {
    for tag in event.tags.iter().filter(|tag| tag_name_is(tag, "e")) {
        if tag.get(3).is_some_and(|marker| marker == "root") {
            push_event_tag(
                refs,
                EventReferenceKind::ReplyRoot,
                Some(tag.as_slice()),
                EventReferenceSource::E,
            );
        }
        if tag.get(3).is_some_and(|marker| marker == "reply") {
            push_event_tag(
                refs,
                EventReferenceKind::ReplyParent,
                Some(tag.as_slice()),
                EventReferenceSource::E,
            );
        }
    }
}

fn push_missing_reply_fallbacks(refs: &mut Vec<EventReference>, event: &NostrEvent) {
    if !refs
        .iter()
        .any(|item| item.kind == EventReferenceKind::ReplyRoot)
    {
        push(
            refs,
            EventReferenceKind::ReplyRoot,
            reply_root(event).as_ref(),
            EventReferenceSource::E,
            Vec::new(),
        );
    }
    if !refs
        .iter()
        .any(|item| item.kind == EventReferenceKind::ReplyParent)
    {
        push(
            refs,
            EventReferenceKind::ReplyParent,
            reply_parent(event).as_ref(),
            EventReferenceSource::E,
            Vec::new(),
        );
    }
}

fn nostr_event_references(content: &str) -> Vec<EventReference> {
    nostr_entities(content)
        .into_iter()
        .filter_map(|entity| match decode_nip19(&entity) {
            Some(NostrEntity::Note(id)) if is_event_id(&id) => {
                Some(content_ref(id, Vec::new(), None))
            }
            Some(NostrEntity::Nevent(pointer)) if is_event_id(&pointer.id) => Some(content_ref(
                pointer.id,
                normalized_relays(pointer.relays.unwrap_or_default()),
                pointer.author,
            )),
            _ => None,
        })
        .collect()
}

fn event_entity_id(entity: &str) -> Option<String> {
    match decode_nip19(entity) {
        Some(NostrEntity::Note(id)) if is_event_id(&id) => Some(id),
        Some(NostrEntity::Nevent(pointer)) if is_event_id(&pointer.id) => Some(pointer.id),
        _ => None,
    }
}

fn content_ref(id: String, relays: Vec<String>, author_pubkey: Option<String>) -> EventReference {
    EventReference {
        kind: EventReferenceKind::NostrEvent,
        id,
        relays,
        author_pubkey,
        marker: None,
        source: EventReferenceSource::Content,
    }
}
